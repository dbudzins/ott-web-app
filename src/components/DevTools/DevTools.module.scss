@use '../../styles/variables';

.container {
  padding: 16px;
}

.button {
  position: fixed;
  bottom: 0;
  left: 0;
  display: flex;
  align-items: center;

  min-width: auto;

  margin-bottom: 8px;
  margin-left: 8px;
}

.dropdown {
  > div {
    min-width: 200px;
  }
}