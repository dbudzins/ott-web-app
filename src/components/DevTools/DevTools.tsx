import { useMemo, useState } from 'react';

import styles from './DevTools.module.scss';

import Dropdown from '#src/components/Dropdown/Dropdown';
import IconButton from '#src/components/IconButton/IconButton';
import Sidebar from '#src/components/Sidebar/Sidebar';
import Settings from '#src/icons/Settings';
import { getStoredConfig, setStoredConfig } from '#src/utils/configOverride';
import { CONFIG_DEFAULT_SOURCE, INCLUDE_CONFIGS } from '#src/env';
import { logDev } from '#src/utils/common';
import { useConfigStore } from '#src/stores/ConfigStore';

// Dynamic value doesn't work for some reason
const getConfigs = () => {
  switch (INCLUDE_CONFIGS) {
    case 'test':
      return import.meta.glob(`../../../configs/test/*.json`);
    case 'prod':
      return import.meta.glob(`../../../configs/prod/*.json`);
    default:
      return [];
  }
};

/**
 * avod-subs--config.json -> Blender Avod Subs
 *
 * We transform the stored config names into readable ones to show in a Select control
 *
 * 1. We remove '--config' part
 * 2. We split words by '-' and add uppercase to the first word's letter
 * 3. We add siteName to the config
 * */
const getReadableConfigName = (siteName: string | undefined, configName: string | undefined | null) => {
  if (!configName) {
    return '';
  }

  const name = configName
    ?.split('--')[0]
    .split('-')
    .map((str) => str.charAt(0).toUpperCase() + str.slice(1))
    .join(' ');

  return siteName ? `${siteName} ${name}` : name;
};

/**
 * Blender Avod Subs  -> avod-subs--config.json
 *
 * 1. We remove siteName from the config
 * 2. We combine words with '-' looking at word breaks
 * 3. We add '--config' part
 *
 */
const getSystemConfigName = (siteName: string | undefined, configName: string | undefined) =>
  configName
    ?.replace(siteName || '', '')
    .trim()
    .toLowerCase()
    .split(' ')
    .join('-') + '--config';

const getConfigOptions = (siteName: string | undefined) =>
  Object.entries(getConfigs()).reduce<string[]>((acc, [pth]) => {
    let configName = '';

    try {
      const parts = pth.split('/');
      const fileName = parts.pop();
      configName = getReadableConfigName(siteName, fileName);
    } catch (err: unknown) {
      logDev('Config file could not be processed: ' + err);
    }

    return configName ? [...acc, configName] : acc;
  }, []);

const DevTools = () => {
  const [showConfigs, setShowConfigs] = useState(false);
  const siteName = useConfigStore((s) => s.config.siteName);

  const configOptions = useMemo(() => getConfigOptions(siteName), [siteName]);
  const selectedConfig = getReadableConfigName(siteName, getStoredConfig() || CONFIG_DEFAULT_SOURCE) || '';
  const defaultLabel = !configOptions.includes(selectedConfig || '') ? getReadableConfigName(siteName, selectedConfig) : undefined;

  const onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setStoredConfig(getSystemConfigName(siteName, event.target.value));
    window.location.reload();
  };

  return (
    <div>
      <Sidebar isOpen={showConfigs} onClose={() => setShowConfigs(false)}>
        <div className={styles.container}>
          <Dropdown
            label="Select config"
            required
            className={styles.dropdown}
            size="small"
            options={configOptions}
            defaultLabel={defaultLabel}
            name="config-select"
            value={selectedConfig}
            onChange={onChange}
          />
        </div>
      </Sidebar>
      <IconButton className={styles.button} aria-label="show-configs-switcher" onClick={() => setShowConfigs(true)}>
        <Settings />
      </IconButton>
    </div>
  );
};

export default DevTools;
