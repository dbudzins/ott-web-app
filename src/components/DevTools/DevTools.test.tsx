import React from 'react';

import DevTools from './DevTools';

import { renderWithRouter } from '#test/testUtils';

describe('<ConfigSelect>', () => {
  test('renders and matches snapshot', () => {
    const { container } = renderWithRouter(<DevTools />);

    expect(container).toMatchSnapshot();
  });
});
