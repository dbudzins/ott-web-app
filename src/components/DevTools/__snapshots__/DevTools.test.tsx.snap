// Vitest Snapshot v1

exports[`<ConfigSelect> > renders and matches snapshot 1`] = `
<div>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="sidebar"
    >
      <div
        class="heading"
      >
        <div
          aria-label="close_menu"
          class="iconButton"
          role="button"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            class="icon"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </g>
          </svg>
        </div>
      </div>
      <nav
        class="group"
      >
        <div
          class="container"
        >
          <div
            class="container dropdown"
          >
            <label
              class="label"
              for="1235"
            >
              Select config
            </label>
            <div
              class="dropdown"
            >
              <select
                aria-required="true"
                class="select"
                id="1235"
                name="config-select"
              >
                <option
                  class="option"
                  value="Accounts"
                >
                  Accounts
                </option>
                <option
                  class="option"
                  value="Blender"
                >
                  Blender
                </option>
                <option
                  class="option"
                  value="Default"
                >
                  Default
                </option>
                <option
                  class="option"
                  value="No Cleeng"
                >
                  No Cleeng
                </option>
                <option
                  class="option"
                  value="Subscription"
                >
                  Subscription
                </option>
                <option
                  class="option"
                  value="Watchlists"
                >
                  Watchlists
                </option>
              </select>
            </div>
          </div>
        </div>
      </nav>
    </div>
    <div
      aria-label="show-configs-switcher"
      class="iconButton button"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="icon"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="m21.49 5.46-3 3-2.97-2.92 3-3A5.7 5.7 0 0 0 16.17 2a5.84 5.84 0 0 0-5.84 5.83 5.77 5.77 0 0 0 .35 2l-7.95 7.9a2.5 2.5 0 0 0 3.54 3.54l7.94-7.95a5.77 5.77 0 0 0 2 .35A5.84 5.84 0 0 0 22 7.83a5.7 5.7 0 0 0-.51-2.37Zm-17 15.31a1.25 1.25 0 1 1 1.25-1.25 1.25 1.25 0 0 1-1.26 1.25Z"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
